usr/src/man/man1/split.1
.\"
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\"
.\" The Institute of Electrical and Electronics Engineers and The Open
.\" Group, have given us permission to reprint portions of their
.\" documentation.
.\"
.\" In the following statement, the phrase ``this text'' refers to portions
.\" of the system documentation.
.\"
.\" Portions of this text are reprinted and reproduced in electronic form
.\" in the SunOS Reference Manual, from IEEE Std 1003.1, 2004 Edition,
.\" Standard for Information Technology -- Portable Operating System
.\" Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group.  In the event of any discrepancy
.\" between these versions and the original IEEE and The Open Group
.\" Standard, the original IEEE and The Open Group Standard is the referee
.\" document.  The original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"
.\" This notice shall appear on any product containing this material.
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright 1989 AT&T
.\" Portions Copyright (c) 1992, X/Open Company Limited  All Rights Reserved
.\" Copyright (c) 1999, Sun Microsystems, Inc.  All Rights Reserved
.\"
.TH SPLIT 1 "Apr 16, 1999"
.SH NAME
split \- split a file into pieces
.SH SYNOPSIS
.LP
.nf
\fBsplit\fR [\fB-linecount\fR | \fB-l\fR \fIlinecount\fR] [\fB-a\fR \fIsuffixlength\fR]
     [\fIfile\fR [\fIname\fR]]
.fi

.LP
.nf
\fBsplit\fR [\fB-b\fR \fIn\fR | \fIn\fRk | \fIn\fRm] [\fB-a\fR \fIsuffixlength\fR] [\fIfile\fR [\fIname\fR]]
.fi

.SH DESCRIPTION
.sp
.LP
The \fBsplit\fR utility reads \fIfile\fR and writes it in \fIlinecount\fR-line
pieces into a set of output-files. The name of the first output-file is
\fIname\fR with \fBaa\fR appended, and so on lexicographically, up to \fBzz\fR
(a maximum of 676 files). The maximum length of \fIname\fR is 2 characters less
than the maximum filename length allowed by the filesystem. See
\fBstatvfs\fR(2). If no output name is given, \fBx\fR is used as the default
(output-files will be called \fBxaa\fR, \fBxab\fR, and so forth).
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB\(mi\fR\fIlinecount\fR \fB|\fR \fB-l\fR \fIlinecount\fR\fR
.ad
.sp .6
.RS 4n
Number of lines in each piece. Defaults to \fB1000\fR lines.
.RE

.sp
.ne 2
.na
\fB\fB-a\fR \fIsuffixlength\fR\fR
.ad
.sp .6
.RS 4n
Uses \fIsuffixlength\fR letters to form the suffix portion of the filenames of
the split file. If \fB-a\fR is not specified, the default suffix length is
\fB2\fR. If the sum of the \fIname\fR operand and the \fIsuffixlength\fR
option-argument would create a filename exceeding \fINAME_MAX\fR bytes, an
error will result; \fBsplit\fR will exit with a diagnostic message and no files
will be created.
.RE

.sp
.ne 2
.na
\fB\fB-b\fR \fIn\fR\fR
.ad
.sp .6
.RS 4n
Splits a file into pieces \fIn\fR bytes in size.
.RE

.sp
.ne 2
.na
\fB\fB-b\fR \fIn\fR\fBk\fR\fR
.ad
.sp .6
.RS 4n
Splits a file into pieces \fIn\fR\fB*1024\fR bytes in size.
.RE

.sp
.ne 2
.na
\fB\fB-b\fR \fIn\fR\fBm\fR\fR
.ad
.sp .6
.RS 4n
Splits a file into pieces \fIn\fR\fB*1\|048\|576\fR bytes in size.
.RE

.SH OPERANDS
.sp
.LP
The following operands are supported:
.sp
.ne 2
.na
\fB\fIfile\fR\fR
.ad
.RS 8n
The path name of the ordinary file to be split. If no input file is given or
\fIfile\fR is \fB\(mi\fR, the standard input will be used.
.RE

.sp
.ne 2
.na
\fB\fIname\fR\fR
.ad
.RS 8n
The prefix to be used for each of the files resulting from the \fBsplit\fR
operation. If no \fIname\fR argument is given, \fBx\fR will be used as the
prefix of the output files. The combined length of the basename of \fIprefix\fR
and \fIsuffixlength\fR cannot exceed \fINAME_MAX\fR bytes. See OPTIONS.
.RE

.SH USAGE
.sp
.LP
See \fBlargefile\fR(7) for the description of the behavior of \fBsplit\fR when
encountering files greater than or equal to 2 Gbyte ( 2^31 bytes).
.SH ENVIRONMENT VARIABLES
.sp
.LP
See \fBenviron\fR(7) for descriptions of the following environment variables
that affect the execution of \fBsplit\fR: \fBLANG\fR, \fBLC_ALL\fR,
\fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR\fR
.ad
.RS 6n
Successful completion.
.RE

.sp
.ne 2
.na
\fB\fB>0\fR\fR
.ad
.RS 6n
An error occurred.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(7) for descriptions of the following attributes:
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
CSI	Enabled
_
Interface Stability	Committed
_
Standard	See  \fBstandards\fR(7).
.TE

.SH SEE ALSO
.sp
.LP
.BR csplit (1),
.BR statvfs (2),
.BR attributes (7),
.BR environ (7),
.BR largefile (7),
.BR standards (7)


usr/src/cmd/split/split.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */

/*
 * Copyright 1999 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved	*/

#pragma ident	"%Z%%M%	%I%	%E% SMI"

#include <stdio.h>
#include <sys/types.h>
#include <sys/statvfs.h>
#include <locale.h>
#include <malloc.h>
#include <string.h>
#include <sys/param.h>
#include <stdlib.h>
#include <wchar.h>
#include <widec.h>
#include <ctype.h>
#include <errno.h>


#define	DEFAULT_LINES	1000
#define	ONE_K		1024
#define	ONE_M		ONE_K*ONE_K
#ifndef	TRUE
#define	TRUE		1
#define	FALSE		0
#endif


static	void	Usage();
static	void	next_file_name();


static	char	*progname;
static	int	suffix_length = 2;

int
main(int argc, char **argv)
{
	long long	line_count = 0;
	long long	byte_count = 0;
	long long	out;
	char	*fname = NULL;
	char	head[MAXPATHLEN];
	char	*output_file_name;
	char	*tail;
	char	*last;
	FILE	*in_file = NULL;
	FILE	*out_file = (FILE *)NULL;
	int	i;
	int	c;
	wint_t	wc;
	int	output_file_open;
	struct	statvfs stbuf;
	int	opt;
	int	non_standard_line_count = FALSE;


	(void) setlocale(LC_ALL, "");
#if !defined(TEXT_DOMAIN)	/* Should be defined by cc -D */
#define	TEXT_DOMAIN "SYS_TEST"	/* Use this only if it weren't */
#endif
	(void) textdomain(TEXT_DOMAIN);

	progname = argv[0];

	/* check for explicit stdin "-" option */
	for (i = 1; i < argc; i++) {
		if (strcmp(argv[i], "-") == 0) {
			in_file = stdin;
			while (i < argc) {
				argv[i] = argv[i + 1];

				/* a "-" before "--" is an error */
				if ((argv[i] != NULL) &&
				    (strcmp(argv[i], "--") == 0)) {
					Usage();
				}
				i++;
			}
			argc--;
		}
	}

	/* check for non-standard "-line-count" option */
	for (i = 1; i < argc; i++) {
		if (strcmp(argv[i], "--") == 0)
			break;

		if ((argv[i][0] == '-') && isdigit(argv[i][1])) {
			if (strlen(&argv[i][1]) !=
			    strspn(&argv[i][1], "0123456789")) {
				(void) fprintf(stderr, gettext(
				    "%s: Badly formed number\n"), progname);
				Usage();
			}

			line_count = (long long) strtoll(&argv[i][1],
			    (char **)NULL, 10);

			non_standard_line_count = TRUE;
			while (i < argc) {
				argv[i] = argv[i + 1];
				i++;
			}
			argc--;
		}
	}

	/* get options */
	while ((opt = getopt(argc, argv, "a:b:l:")) != EOF) {
		switch (opt) {
		case 'a':
			if (strcmp(optarg, "--") == 0) {
				Usage();
			}
			suffix_length = (long long) strtoll(optarg,
					(char **)NULL, 10);
			if (suffix_length <= 0) {
				(void) fprintf(stderr, gettext(
				    "%s: Invalid \"-a %s\" option\n"),
				    progname, optarg);
				Usage();
			}
			break;

		case 'b':
			if (strcmp(optarg, "--") == 0) {
				Usage();
			}
			byte_count = (long long) strtoll(optarg,
					(char **)NULL, 10);
			if (*(optarg + strspn(optarg, "0123456789")) == 'k')
				byte_count *= ONE_K;
			if (*(optarg + strspn(optarg, "0123456789")) == 'm')
				byte_count *= ONE_M;
			break;

		case 'l':
			if (strcmp(optarg, "--") == 0) {
				Usage();
			}
			if (non_standard_line_count == TRUE) {
				Usage();
			}
			line_count = (long long) strtoll(optarg,
					(char **)NULL, 10);
			break;

		default:
			Usage();
		}
	}

	/* get input file */
	if ((in_file == NULL) && (optind < argc)) {
		if ((in_file = fopen(argv[optind++], "r")) == NULL) {
			(void) perror("split");
			return (1);
		}
	}
	if (in_file == NULL) {
		in_file = stdin;
	}

	/* get output file name */
	if (optind < argc) {
		output_file_name = argv[optind];
		if ((tail = strrchr(output_file_name, '/')) == NULL) {
			tail = output_file_name;
			(void) getcwd(head, sizeof (head));
		} else {
			tail++;
			(void) strcpy(head, output_file_name);
			last = strrchr(head, '/');
			*++last = '\0';
		}

		if (statvfs(head, &stbuf) < 0) {
			perror(head);
			return (1);
		}

		if (strlen(tail) > (stbuf.f_namemax - suffix_length)) {
			(void) fprintf(stderr, gettext(
			    "%s: More than %d characters in file name\n"),
			    progname, stbuf.f_namemax - suffix_length);
			Usage();
		}
	} else
		output_file_name = "x";

	/* check options */
	if (((int)strlen(output_file_name) + suffix_length) > FILENAME_MAX) {
		(void) fprintf(stderr, gettext(
		"%s: Output file name too long\n"), progname);
		return (1);
	}

	if (line_count && byte_count) {
		    Usage();
	}

	/* use default line count if none specified */
	if (line_count == 0) {
		line_count = DEFAULT_LINES;
	}

	/*
	 * allocate buffer for the filenames we'll construct; it must be
	 * big enough to hold the name in 'output_file_name' + an n-char
	 * suffix + NULL terminator
	 */
	if ((fname = (char *)malloc(strlen(output_file_name) +
	    suffix_length + 1)) == NULL) {
		(void) perror("split");
		return (1);
	}

	/* build first output file name */
	for (i = 0; output_file_name[i]; i++) {
		fname[i] = output_file_name[i];
	}
	while (i < (int)strlen(output_file_name) + suffix_length) {
		fname[i++] = 'a';
	}
	if (suffix_length)
		fname[i - 1] = 'a' - 1;
	fname[i] = '\0';

	for (; ; ) {
	    output_file_open = FALSE;
	    if (byte_count) {
		for (out = 0; out < byte_count; out++) {
		    errno = 0;
		    c = getc(in_file);
		    if (c == EOF) {
			if (errno != 0) {
			    int lerrno = errno;
			    (void) fprintf(stderr, gettext(
				"%s: Read error at file offset %lld: %s, "
				"aborting split\n"),
				    progname, ftello(in_file),
				    strerror(lerrno));
			    if (output_file_open == TRUE)
				(void) fclose(out_file);
			    free(fname);
			    return (1);
			}
			if (output_file_open == TRUE)
			    (void) fclose(out_file);
			free(fname);
			return (0);
		    }
		    if (output_file_open == FALSE) {
			next_file_name(fname);
			if ((out_file = fopen(fname, "w")) == NULL) {
			    (void) perror("split");
			    free(fname);
			    return (1);
			}
			output_file_open = TRUE;
		    }
		    if (putc(c, out_file) == EOF) {
			perror("split");
			if (output_file_open == TRUE)
			    (void) fclose(out_file);
			free(fname);
			return (1);
		    }
		}
	    } else {
		for (out = 0; out < line_count; out++) {
		    do {
			errno = 0;
			wc = getwc(in_file);
			if (wc == WEOF) {
			    if (errno != 0) {
				if (errno == EILSEQ) {
				    (void) fprintf(stderr, gettext(
					"%s: Invalid multibyte sequence "
					"encountered at file offset %lld, "
					"aborting split\n"),
					    progname, ftello(in_file));
				} else {
				    (void) perror("split");
				}
				if (output_file_open == TRUE)
				    (void) fclose(out_file);
				free(fname);
				return (1);
			    }
			    if (output_file_open == TRUE)
				(void) fclose(out_file);
			    free(fname);
			    return (0);
			}
			if (output_file_open == FALSE) {
			    next_file_name(fname);
			    if ((out_file = fopen(fname, "w")) == NULL) {
				(void) perror("split");
				free(fname);
				return (1);
			    }
			    output_file_open = TRUE;
			}
			if (putwc(wc, out_file) == WEOF) {
			    (void) perror("split");
			    if (output_file_open == TRUE)
				(void) fclose(out_file);
			    free(fname);
			    return (1);
			}
		    } while (wc != '\n');
		}
	    }
	    if (output_file_open == TRUE)
		(void) fclose(out_file);
	}

	/*NOTREACHED*/
}


static	void
next_file_name(char *name)
{
	int	i;

	i = strlen(name) - 1;

	while (i >= (int)(strlen(name) - suffix_length)) {
		if (++name[i] <= 'z')
			return;
		name[i--] = 'a';
	}
	(void) fprintf(stderr, gettext(
		"%s: Exhausted output file names, aborting split\n"),
		progname);
	exit(1);
}


static	void
Usage()
{
	(void) fprintf(stderr, gettext(
		"Usage: %s [-l #] [-a #] [file [name]]\n"
		"       %s [-b #[k|m]] [-a #] [file [name]]\n"
		"       %s [-#] [-a #] [file [name]]\n"),
		progname, progname, progname);
	exit(1);
}
